substrate/frame/asset-conversion/src/lib.rs

```diff
@@ -765,22 +756,119 @@ pub mod pallet {
 			}
 
 			Self::validate_swap_path(&path)?;
+			let path = Self::balance_path_from_amount_out(amount_out, path)?;
 
-			let amounts = Self::get_amounts_in(&amount_out, &path)?;
-			let amount_in =
-				*amounts.first().defensive_ok_or("get_amounts_in() returned an empty result")?;
-
+			let amount_in = path.first().map(|(_, a)| *a).ok_or(Error::<T>::InvalidPath)?;
 			if let Some(amount_in_max) = amount_in_max {
 				ensure!(
 					amount_in <= amount_in_max,
 					Error::<T>::ProvidedMaximumNotSufficientForSwap
 				);
 			}
 
-			Self::do_swap(sender, &amounts, path, send_to, keep_alive)?;
+			Self::swap(&sender, &path, &send_to, keep_alive)?;
+
+			Self::deposit_event(Event::SwapExecuted { who: sender, send_to, path });
+
 			Ok(amount_in)
 		}
 
+		/// Swap exactly `credit_in` of asset `path[0]` for asset `path[last]`.  If `amount_out_min`
+		/// is provided and the swap can't achieve at least this amount, an error is returned.
+		///
+		/// On a successful swap, the function returns the `credit_out` of `path[last]` obtained
+		/// from the `credit_in`. On failure, it returns an `Err` containing the original
+		/// `credit_in` and the associated error code.
+		///
+		/// WARNING: This may return an error after a partial storage mutation. It should be used
+		/// only inside a transactional storage context and an Err result must imply a storage
+		/// rollback.
+		pub(crate) fn do_swap_exact_credit_tokens_for_tokens(
+			path: BoundedVec<T::MultiAssetId, T::MaxSwapPathLength>,
+			credit_in: Credit<T>,
+			amount_out_min: Option<T::AssetBalance>,
+		) -> Result<Credit<T>, (Credit<T>, DispatchError)> {
+			let amount_in = match credit_in.peek() {
+				Ok(a) => a,
+				Err(e) => return Err((credit_in, e)),
+			};
+			let inspect_path = |credit_asset| {
```

Comment:
helps to avoid matching a non clone credit object for returning it in case of error 
alternative solution could be - https://gist.github.com/muharem/24376bcaa4ab19c3e7e0974cbc063648